systemdev/testing-locally.rst

```diff
@@ -3,18 +3,207 @@ Making changes and testing locally
+
+We'll examine the use of crossbuilder and builds on the device using `address-book-app <https://github.com/ubports/address-book-app>`__ (the Contacts application) as an example.
 
 We only recommend developing packages using a device with Ubuntu Touch installed from the devel channel. This ensures that you are testing your changes against the most current state of the Ubuntu Touch code.
 
-.. warning::
+Preparing a package for a build
+-------------------------------
 
-    Installing packages has a risk of damaging the software on your device, rendering it unusable. If this happens, you can :doc:`reinstall Ubuntu Touch </userguide/install>`.
+``sbuild`` or ``crossbuilder`` need to be invoked from a debianized package source tree (i.e. the package sources with a ``debian`` subdirectory), UBports packages mostly consists of git repositories containing native or non-native packages.  Native packages can be built directly and the following script can be used in order to prepare a non-native package for a build::
+
+    #!/bin/sh
+
+    PATH=/bin:/usr/bin
+
+    die() {
+        if [ $# -gt 0 ]; then
+            printf "%s\n" "$1" >&2
+        fi
+        exit 1
+    }
+
+    if [ ! -d "./debian" ]; then
+        die "not in a project directory"
+    fi
+
+    if [ -f "./debian/ubports.source_location" ]; then
+
+        {
+            read -r src_url && \
+            read -r src_filename
+        } < "./debian/ubports.source_location" || \
+            die "failed to parse ubports.source_location"
+        case ${src_url} in
+        http://*|https://*|ftp://*)
+            ;;
+        *)
+            die "invalid url: \"${src_url}\""
+            ;;
+        esac
+        src_filename="$(basename "${src_filename}")"
+
+        wget -O "../${src_filename}" "${src_url}" || \
+            die "failed to download source archive"
+    fi
+
+
+Building packages in a chroot using sbuild
+------------------------------------------
+
+`sbuild <https://tracker.debian.org/pkg/sbuild>`__ is a tool for building Debian packages from source in an isolated environment using a chroot created by `schroot <https://wiki.debian.org/Schroot>`__. It closely resembles the package build process on the UBports CI system by using an isolated build environment with a minimal set of pre-installed packages. This detects any missing build dependencies and sbuild will also detect problems with the resulting packages by running `lintian <https://lintian.debian.org/manual/lintian.html>`__
+
+Prerequisites
+^^^^^^^^^^^^^
+
+A host system running either Debian 11 (Bullseye) or later or Ubuntu 20.04 (Focal Fossa) or later is required. It is possible to set up such a system in a LXD container or VM (not recommended due to performance) on another distribution or operating system.
```

Comment:
> I'm still struggling with the LXD stuff here. Am I right that the part from 'It is possible to' until the lxc commandline in line 71 ... all of this is optional right? Could we pull this out and put it somewhere as a standalone paragraph or section rather than in the middle of the Prerequisites section?

It is not needed if your're running Debian or Ubuntu. If someone is using another distribution which does not have Debian devscripts, debootstrap and sbuild prepackaged, LXD which can be installed via snap on many distributions is an easy and performant way to get an Ubuntu or Debian environment going where one could also install and test the built debs. I've tried to clarify that in the prerequisites section.